tests/test_docker-compose.sh

```diff
@@ -0,0 +1,660 @@
+    if [ $rc -ne 0 ]; then
+        echo "ArtifactRollback failed (exit code $rc), logs follow:"
+        cat "${WORKDIR}/docker-compose.log"
+        return $rc
+    fi
+
+    "${SRCDIR}/docker-compose" Cleanup "${WORKDIR}/artifact" >> "${WORKDIR}/docker-compose.log" 2>&1 || rc=$?
+    if [ $rc -ne 0 ]; then
+        echo "Second Cleanup failed (exit code $rc), logs follow:"
+        cat "${WORKDIR}/docker-compose.log"
+        return $rc
+    fi
+
+    cat << EOF | diff -u - "$CMDLINE_LOGGER_LOG_FILE" || rc=$?
+docker image load
+docker image load
+docker images --format {{json .ID}} some/lighttpd:latest
+docker images --format {{json .ID}} bad/php:oldest
+docker-compose --project-name test-comp up -d
+docker image load
+docker images --format {{json .ID}} bad/php:worst
+docker-compose --project-name test-comp down
+docker-compose --project-name test-comp up -d
+docker-compose --project-name test-comp down
+docker-compose --project-name test-comp up -d
+docker rmi bad/php:worst
+EOF
+    if [ $rc -ne 0 ]; then
+        echo "Unexpected commands executed (see the above diff), logs follow:"
+        cat "${WORKDIR}/docker-compose.log"
+        return $rc
+    fi
+
+    if [ -d "${PERSISTENT_DIR}/new" ]; then
+        echo "'new' directory exists after rollback"
+        return 1
+    fi
+    if ! [ -d "${PERSISTENT_DIR}/current" ]; then
+        echo "'current' directory doesn't exist after rollback"
+        return 1
+    fi
+    if [ -d "${PERSISTENT_DIR}/cleanup" ]; then
+        echo "'cleanup' directory exists after rollback"
+        return 1
+    fi
+
+    return $rc
+}
+tests+=(test_artifact_install_commit_install_rollback_cleanup)
+
+test_artifact_install_load_fail_rollback_cleanup() {
+    local rc=0
+
+    prepare_config
+    prepare_expected_file_tree
+    prepare_docker_mock
+
+    # We need a 'docker load' failure, ideally for the second image which is the
+    # more tricky case.
+    cat << EOF > "${WORKDIR}/bin/docker"
+#!/bin/bash
+case "\$1" in
+     --version|version)
+        exit 0
+        ;;
+     compose)
+        exit 1
+        ;;
+     images)
+        if [ -f "${WORKDIR}/artifact/tmp/first_img_query_done" ]; then
+            echo "\$(basename \$0) \$@" >> "$CMDLINE_LOGGER_LOG_FILE"
+            exit 1
+        else
+            touch "${WORKDIR}/artifact/tmp/first_img_query_done"
+            cat /proc/sys/kernel/random/uuid
+        fi
+        ;;
+     image) # docker image load < /some/img.tar
+        if [ -f "${WORKDIR}/artifact/tmp/first_load_done" ]; then
+            echo "\$(basename \$0) \$@" >> "$CMDLINE_LOGGER_LOG_FILE"
+            exit 1
+        else
+            touch "${WORKDIR}/artifact/tmp/first_load_done"
+        fi
+        ;;
+esac
+echo "\$(basename \$0) \$@" >> "$CMDLINE_LOGGER_LOG_FILE"
+exit 0
+EOF
+
+    "${SRCDIR}/docker-compose" ArtifactInstall "${WORKDIR}/artifact" > "${WORKDIR}/docker-compose.log" 2>&1 || rc=$?
+    if [ $rc -eq 0 ]; then
```

Comment:
For consistency with the rest of the test suite, why not `-ne 1` ? It is cleaner to check for an specific expectation and treat the rest as errors.

Reply:
Yup. I didn't want to worry about specific error exit codes, but you're right.